### ISSUE TITLE
0.91 - trouble getting MAX31855 thermocouple chip running

### ISSUE THREAD

**Issue by simonkuehling** [repetier/Repetier-Firmware#381]
Hi Roland,

i am trying to wire a MAX31855 breakout board up to a RUMBA board for some high-temp hotend developments. But i have trouble getting the communication to work. Or more precisely: It seems that i cannot get an SPI clock signal on the configured SCK_PIN...

I tried a simple test program on the very same RUMBA board to verify that the MAX31855 is working on the pins i chose. It just won't run with repetier firmware.

These are the relevant configurations i made:
```
#define EXT0_TEMPSENSOR_TYPE 102
#define SUPPORT_MAX31855
#define TEMP_0_PIN         33       // unused Z-MIN endstop pin as CS (chip select)
#define SCK_PIN            52
#define MISO_PIN           50
```

Is there any additional setting i missed to enable SPI communications?
Did i get the chip select configuration right by defining it as ```TEMP_0_PIN```?
Thanks for any hint!

**Comment by repetier** [comment]
No, init is called for each test, see in Extruder.cpp the read function

#ifdef SUPPORT_MAX31855
int16_t read_max31855(uint8_t ss_pin)
{
    uint32_t data = 0;
    int16_t temperature;
    HAL::spiInit(1);
    HAL::digitalWrite(ss_pin, 0);  // enable TT_MAX31855
    HAL::delayMicroseconds(1);    // ensure 100ns delay - a bit extra is fine

    for (unsigned short byte = 0; byte < 4; byte++)
    {
        data <<= 8;
        data |= HAL::spiReceive();
    }

    HAL::digitalWrite(ss_pin, 1);  // disable TT_MAX31855

    //Process temp
    if (data & 0x00010000)
        return 20000; //Some form of error.
    else
    {
        data = data >> 18;
        temperature = data & 0x00001FFF;

        if (data & 0x00002000)
        {
            data = ~data;
            temperature = -1 * ((data & 0x00001FFF) + 1);
        }
    }
    return temperature;
}
#endif

**Comment by simonkuehling** [comment]
@repetier, @pkmcnc, glad that i am not alone with this issues - i just assumed this feature is tested/working as it is in there for quite a while already..

i am really eager to get this running - so Roland, whatever i can try or test, i have all components on my desk at the moment  :-)

**Comment by pkmcnc** [comment]
I also have MAX6675 with a spare thermocouple and ramps at hand. Ready for testing. 
I should have created a ticket a few months back when I found this problem...

@simonkuehling Did you try another firmware? I didn't )

**Comment by simonkuehling** [comment]
@pkmcnc, i only tried repetier firmware so far as the thermocouple will need to run with this platform. But as mentioned, i first tested the general RUMBA+MAX31855 combination with a quick sketch i uploaded to the ATMEGA instead of the firmware.

Does repetier firmware use software or hardware SPI, btw? I only tried software SPI in the sketch so far, which worked flawlessly

**Comment by repetier** [comment]
Maybe thats the problem. It is definitvely hardware SPI as it uses the hardwarse spi routines for the sd card for communication. So MOSI, MISO and SCK must match the hardware pins. Do they match?

**Comment by simonkuehling** [comment]
yes, I left MISO, MOSI (unused) and SCK in the default hardware spi pin configuration of the atmega2560.

I still did not test the max31855 with hrdware SPI yet, but from the adafruit library for my breakout board i see that it needs to be queried in SPI mode "0".  Is the mode setting a global config in the firmware somewhere or is it set on each ```spiInit()```?
In which mode does the SD communication run?

**Comment by simonkuehling** [comment]
Another thing might be the SPI clock rate. From the datasheet i see a max of 5Mhz for the MAX31855. Which would be fine with a clock divider of 4 (=4Mhz) on the atmega2560.
Is SD communication also running on this rate?

**Comment by simonkuehling** [comment]
Is the ```spiRate``` argument for ```spiInit()``` the exact clock divider? I did not fully understand the register setup in ```spiInit``` yet...

If this is the divider, we may have the problem in ```HAL::spiInit(1)``` from ```read_max31855()```...

**Comment by repetier** [comment]
Se HAL.h spiInit

    static inline void spiInit(uint8_t spiRate)
    {
         uint8_t r = 0;
         for (uint8_t b = 2; spiRate > b && r < 6; b <<= 1, r++);

        SET_OUTPUT(SS);
        WRITE(SS,HIGH);
        SET_OUTPUT(SCK);
        SET_OUTPUT(MOSI_PIN);
        SET_INPUT(MISO_PIN);
#ifdef	PRR
        PRR &= ~(1<<PRSPI);
#elif defined PRR0
        PRR0 &= ~(1<<PRSPI);
#endif
        // See avr processor documentation
        SPCR = (1 << SPE) | (1 << MSTR) | (r >> 1);
        SPSR = (r & 1 || r == 6 ? 0 : 1) << SPI2X;

    }

the parameter in spiInit is the speed. 
spiRate 1 => r = 1, b = 4 => Clock / 2
SpiRate 2 => r = 0, b = 2 => Clock / 4
if i interpret it correctly.

SPCR also sets SPI mode which is as you see from formula 0.

**Comment by simonkuehling** [comment]
**SUCCESS!**

With
```
HAL::spiInit(2);
```
in ```read_max31855()``` from Extruder.cpp, the MAX31855 is now working as expected!
Shall i make a pull request for this one? ;-)

The same fix should apply for ```read_max6675()``` as the MAX6675 datasheet states a max serial clock frequency of 4.3MHz...

**Comment by repetier** [comment]
No need for a pull request. I'll change the 2 byte directly so they will be online with the next update.

Thanks for trying and updating.

**Comment by pkmcnc** [comment]
MAX6675 still won't work, unfortunately. 
It gives meaningful readings in repetier-host but never updates.
MAX6675 is testing OK it with a simple software.
I found that MAX6675_SS used in pins.h but SS used in extruder.cpp. Corrected it but still nothing.

**Comment by simonkuehling** [comment]
@pkmcnc, i am not sure what the MAX6675_SS pin configuration is used for, but since ```read_max6675()``` and ```read_max31855()``` are called from within the same code block in Extruder.cpp i suggest trying to instead set your TEMP_0_PIN to the one used for activating the chip. At least this is the configuration i am using.
The chip select is a extruder-specific configuration in order to always only activate one thermocouple chip at a time.

I am curious if this is all that is missing :-)

**Comment by pkmcnc** [comment]
@simonkuehling I searched through the code a few times, and also could not find where MAX6675_SS is used. And I already set TEMP_0_PIN to 53 which activates the chip, this is so not obvious BTW. MAX6675_SS and TEMP_0_PIN is basically the same then...
Actually I also tried SD support on/off, no luck.
Can't understand what is different from 31855. I'll try to choose 31855 but use spi read code from 6675.
